library/core/src/num/mod.rs

```diff
@@ -846,37 +849,79 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+    if mem::size_of::<T>() <= mem::size_of::<u32>() && radix == 10 {
+        // The ALU can reorder these adds and do more in parallel
+        // as each mul isn't dependent on the previous answer.
+        let factors = &MULTIPLIER[MULTIPLIER.len() - digits.len()..];
+        let mut idx = 0;
+        if is_positive {
+            // The number is positive
+            for &c in digits {
+                let x = match (c as char).to_digit(radix) {
+                    Some(x) => x,
+                    None => return Err(PIE { kind: InvalidDigit }),
+                };
```

Comment:
These `match`es can all be written a bit shorter like this:
```suggestion
                let x = (c as char).to_digit(radix).ok_or(PIE { kind: InvalidDigit })?;
```

Reply:
I recall we used to do these but in performance sensitive parts using `ok_or` is slower than `match` counterparts, from what I recall.

Reply:
Really? The optimizer should have all information it needs and in my quick tests the generated assembly is identcial. Is there an open issue or do you have a test case?

Reply:
No, that's what I recall but not sure if the same goes for this case, if the assembly is the same then I guess it's good to have it.